File: libgomp/testsuite/libgomp.fortran/loop-transforms/unroll-1.f90
! { dg-additional-options "-fdump-tree-original" }
! { dg-do run }

module test_functions
  contains
  integer function compute_sum() result(sum)
    implicit none

    integer :: i,j

    sum = 0
    !$omp do
    do i = 1,10,3
       !$omp unroll full
       do j = 1,10,3
          sum = sum + 1
       end do
    end do
  end function

  integer function compute_sum2() result(sum)
    implicit none

    integer :: i,j

    sum = 0
    !$omp parallel do reduction(+:sum)
    !$omp unroll partial(2)
    do i = 1,10,3
       do j = 1,10,3
          sum = sum + 1
       end do
    end do
  end function
end module test_functions

program test
  use test_functions
  implicit none

  integer :: result

  result = compute_sum ()
  write (*,*) result
  if (result .ne. 16) then
     call abort
  end if

  result = compute_sum2 ()
  write (*,*) result
  if (result .ne. 16) then
     call abort
  end if
end program

File: libgomp/testsuite/libgomp.c-c++-common/unroll-1.c
/* { dg-additional-options "-Wall -Wno-unknown-pragmas" } */

int
compute_sum1 (void)
{
  int sum = 0;
  int i, j;

  #pragma omp parallel for reduction(+:sum) lastprivate(j)
  #pragma omp unroll partial
  for (i = 3; i < 10; ++i)
    for (j = -2; j < 7; ++j)
      sum++;

  if (j != 7)
    __builtin_abort ();

  return sum;
}

int
compute_sum2 (void)
{
  int sum = 0;
  int i, j;

  #pragma omp parallel for reduction(+:sum) lastprivate(j)
  #pragma omp unroll partial(5)
  for (i = 3; i < 10; ++i)
    for (j = -2; j < 7; ++j)
      sum++;

  if (j != 7)
    __builtin_abort ();

  return sum;
}

int
compute_sum3 (void)
{
  int sum = 0;
  int i, j;

  #pragma omp parallel for reduction(+:sum) lastprivate(j)
  #pragma omp unroll partial(1)
  for (i = 3; i < 10; ++i)
    for (j = -2; j < 7; ++j)
      sum++;

  if (j != 7)
    __builtin_abort ();

  return sum;
}

int
main ()
{
  if (compute_sum1 () != 7 * 9)
    __builtin_abort ();

  if (compute_sum2 () != 7 * 9)
    __builtin_abort ();

  if (compute_sum3 () != 7 * 9)
    __builtin_abort ();

  return 0;
}
